bake/hcl_test.go

```diff
@@ -696,3 +696,52 @@ target "b" {
 	require.Equal(t, ".", *c.Targets[3].Context)
 	require.Equal(t, "b", *c.Targets[3].Target)
 }
+
+func TestCombineHCLAndJSONWithGlobalAttrs(t *testing.T) {
+	c, err := ParseFiles([]File{
+		{
+			Name: "docker-bake.hcl",
+			Data: []byte(`
+variable "ABC" {
+  default = "foo"
+}
+variable "DEF" {
+  default = ""
+}
+group "default" {
+  targets = ["one"]
+}
+target "one" {
+  args = {
+    a = "pre-${ABC}"
+  }
+}
+target "two" {
+  args = {
+    b = "pre-${DEF}"
+  }
+}`),
+		},
+		{
+			Name: "foo.json",
+			Data: []byte(`{"variable": {"DEF": {"default": "bar"}}, "target": { "one": { "args": {"a": "pre-${ABC}-${DEF}"}} } }`),
+		},
+		{
+			Name: "bar.json",
+			Data: []byte(`{"ABC": "ghi", "DEF": "jkl"}`),
```

Comment:
Does it still work when hcl also defines `ABC = "old"` and `foo.json` `"ABC": "old2"`

Reply:
Indeed I tried with attributes in global scope an doesn't seem to work atm (see last commit)

```
=== RUN   TestCombineHCLAndJSONAttrs
    hcl_test.go:790: 
        	Error Trace:	hcl_test.go:790
        	Error:      	Not equal: 
        	            	expected: map[string]string{"a":"pre-ghi-jkl"}
        	            	actual  : map[string]string{"a":"pre-foo-jkl"}
```

I guess it's how merging blocks are handled for attributes in global scope where only the first one is retained atm:

```hcl
ABC = "foo"
variable "DEF" {
  default = ""
}
group "default" {
  targets = ["one"]
}
target "one" {
  args = {
    a = "pre-${ABC}"
  }
}
target "two" {
  args = {
    b = "pre-${DEF}"
  }
}
```

```json
{"ABC": "oof", "variable": {"DEF": {"default": "bar"}}, "target": { "one": { "args": {"a": "pre-${ABC}-${DEF}"}} } }
```

```json
{"ABC": "ghi", "DEF": "jkl"}
```

Reply:
Does it at least work with HCL-only?

Reply:
Same with HCL only (see test in last commit) :disappointed::

```
=== RUN   TestCombineHCLAttrs
    hcl_test.go:801: 
        	Error Trace:	hcl_test.go:801
        	Error:      	Not equal: 
        	            	expected: map[string]string{"a":"pre-ghi-jkl"}
        	            	actual  : map[string]string{"a":"pre-foo-jkl"}
```

Reply:
Looks like we missed a case in https://github.com/docker/buildx/pull/541. Changing this test with a global attr `FOO = "abc"` in first file makes it failed:

```go
func TestHCLMultiFileAttrs(t *testing.T) {
	os.Unsetenv("FOO")
	dt := []byte(`
		FOO = "abc"
		target "app" {
			args = {
				v1 = "pre-${FOO}"
			}
		}
		`)
	dt2 := []byte(`
		FOO="def"
		`)

	c, err := ParseFiles([]File{
		{Data: dt, Name: "c1.hcl"},
		{Data: dt2, Name: "c2.hcl"},
	}, nil)
	require.NoError(t, err)
	require.Equal(t, 1, len(c.Targets))
	require.Equal(t, c.Targets[0].Name, "app")
	require.Equal(t, "pre-def", c.Targets[0].Args["v1"])

	os.Setenv("FOO", "ghi")

	c, err = ParseFiles([]File{
		{Data: dt, Name: "c1.hcl"},
		{Data: dt2, Name: "c2.hcl"},
	}, nil)
	require.NoError(t, err)

	require.Equal(t, 1, len(c.Targets))
	require.Equal(t, c.Targets[0].Name, "app")
	require.Equal(t, "pre-ghi", c.Targets[0].Args["v1"])
}
```

```
=== RUN   TestHCLMultiFileAttrs
    hcl_test.go:488: 
        	Error Trace:	hcl_test.go:488
        	Error:      	Not equal: 
        	            	expected: "pre-def"
        	            	actual  : "pre-abc"
```

Reply:
Ah `JustAttributes` skips repeated attributes: https://github.com/hashicorp/hcl/blob/c7ee8b78101c33b4dfed2641d78cf5e9651eabb8/merged.go#L109-L120